src/ipv6-pinhole-test.cpp

```diff
@@ -0,0 +1,322 @@
+        return dest_addr;
+    }
+    return std::nullopt;
+}
+
+//! Try to open a IPv6 pinhole port using RFC 6887 Port Control Protocol (PCP).
+bool OpenPinhole(const sockaddr_in6 &gateway, const struct in6_addr &my_addr, uint16_t port, uint32_t lifetime, int num_tries = 3)
+{
+    LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "pcp6: Opening pinhole for addr: %s\n", CNetAddr(my_addr).ToStringAddr());
+
+    // Create IPv6 UDP socket.
+    SOCKET sock_fd = socket(AF_INET6, SOCK_DGRAM, IPPROTO_UDP);
+    if (sock_fd == INVALID_SOCKET) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "pcp6: Could not create IPv6 UDP socket: %s\n", NetworkErrorString(WSAGetLastError()));
+        return false;
+    }
+    Sock sock(sock_fd);
+
+    // Make sure that we send from requested address, anything else will be
+    // rejected by a security-conscious router.
+    sockaddr_in6 bind_addr = {};
+    bind_addr.sin6_family = AF_INET6;
+    bind_addr.sin6_addr = my_addr;
+    if (sock.Bind((struct sockaddr*)&bind_addr, sizeof(sockaddr_in6)) != 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "pcp6: Could not bind to address: %s\n", NetworkErrorString(WSAGetLastError()));
+        return false;
+    }
+
+    sockaddr_in6 dest_addr = gateway;
+    dest_addr.sin6_port = htons(PCP_SERVER_PORT);
+    if (sock.Connect((struct sockaddr*)&dest_addr, sizeof(sockaddr_in6)) != 0) {
+        LogPrintLevel(BCLog::NET, BCLog::Level::Warning, "pcp6: Could not bind to connect to gateway: %s\n", NetworkErrorString(WSAGetLastError()));
+        return false;
+    }
+
+    // Build request packet. Make sure the packet is zeroed so that reserved fields are zero
+    // as required by the spec (and not potentially leak data).
+    // Make sure there's space for the request header, MAP specific request
+    // data, and one option header.
+    uint8_t request[PCP_REQUEST_HDR_SIZE + PCP_MAP_REQUEST_SIZE + PCP_OPTION_HDR_SIZE] = {};
+    // Fill in request header, See Figure 2.
+    size_t ofs = 0;
+    request[ofs + 0] = PCP_VERSION;
+    request[ofs + 1] = PCP_OP_MAP_REQUEST;
+    WriteBE32(request + ofs + 4, lifetime);
+    memcpy(request + ofs + 8, &my_addr, ADDR_IPV6_SIZE);
+
+    ofs += PCP_REQUEST_HDR_SIZE;
+
+    // Fill in MAP request packet, See Figure 9.
+    // Randomize mapping nonce (this is repeated in the response, to be able to
+    // correlate requests and responses, and used to authenticate changes to the mapping).
+    // TODO: remember this value when periodically renewing a mapping.
+    GetRandBytes(Span(request + ofs + 0, PCP_MAP_NONCE_SIZE));
+    request[ofs + 12] = PCP_PROTOCOL_TCP;
+    WriteBE16(request + ofs + 16, port);
+    WriteBE16(request + ofs + 18, port);
+    memcpy(request + ofs + 20, &my_addr, ADDR_IPV6_SIZE);
+
+    ofs += PCP_MAP_REQUEST_SIZE;
+
+    // Fill in option header. See Figure 4.
+    // Prefer failure to a different external address mapping than we expect.
+    request[ofs] = PCP_OPTION_PREFER_FAILURE;
```

Comment:
We should explicitly set that we want this option to be mandatory (if we do indeed care that much):

> Its most significant bit indicates if this
> option is mandatory (0) or optional (1) to process.

Reply:
As i understand, mandatory or optional is a property of the opcode. `PCP_OPTION_PREFER_FAILURE` is `0x02`, so It's always mandatory.

We want the address and port that we request, if not possible, it's better to fail fast, instead of having to deal with having to cancel unwanted bindings etc.

Reply:
Maybe I misread it, but I think we're implicitly doing:

```h
PCP_OPTION_MANDATORY = 0x00
... PCP_OPTION_PREFER_FAILURE | PCP_OPTION_MANDATORY
```

Reply:
i don't think that's the case. The mandatory flag is not a user choice, it's part of the RFC definition of the option:
- Figure 14 says "Option Code=2" only.
- Section 19.4 says "The values 0-127 are mandatory to process, and 128-255 are optional to process." This implies these are disjunct ranges.

In any case, we could be more lenient about this if we're willing to accept different addresses (or maybe even ports) and advertise them.

Reply:
I was looking at 7.3:

> Option Code:  8 bits.  Its most significant bit indicates if this
>      option is mandatory (0) or optional (1) to process.

But that's implied by using these integer ranges.


Tend to agree we should drop this option, though we should pay attention to its return value. Otherwise we'd potentially gossip the wrong port (in practice this likely only happens if you have two nodes in the same home).